From 91a5dd9c8365607102e6281cfa2de00c8b001fc7 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Tue, 24 Sep 2024 16:47:19 +0200
Subject: [PATCH 0164/1010] net/page_pool: dovetail: define internal accessors
 to cache size/refill

Some drivers (e.g. mellanox/mx5 do refer to PP_ALLOC_CACHE_*
definitions, we cannot change their definition as we'd like to support
dynamic allocation for oob mode. Revert those changes, introducing
dedicated helpers in replacement.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/net/page_pool.h | 41 +++++++++++++++++++++++------------------
 net/core/page_pool.c    |  6 +++---
 2 files changed, 26 insertions(+), 21 deletions(-)

diff --git a/include/net/page_pool.h b/include/net/page_pool.h
index db929c3640..777633fee0 100644
--- a/include/net/page_pool.h
+++ b/include/net/page_pool.h
@@ -66,24 +66,17 @@
  * cache is already full (or partly full) then the XDP_DROP recycles
  * would have to take a slower code path.
  */
-#ifdef CONFIG_PAGE_POOL_OOB
-#define PP_ALLOC_CACHE_SIZE(__pool)	((__pool)->p.pool_size)
-#define PP_ALLOC_CACHE_REFILL(__pool)	PP_ALLOC_CACHE_SIZE(__pool)
+#define PP_ALLOC_CACHE_SIZE	128
+#define PP_ALLOC_CACHE_REFILL	64
 struct pp_alloc_cache {
 	u32 count;
+#ifdef CONFIG_PAGE_POOL_OOB
 	struct page **cache;
 	hard_spinlock_t oob_lock;
-};
 #else	/* !CONFIG_PAGE_POOL_OOB */
-#define __PP_ALLOC_CACHE_SIZE	128
-#define __PP_ALLOC_CACHE_REFILL	64
-#define PP_ALLOC_CACHE_SIZE(__pool)	({ (void)__pool; __PP_ALLOC_CACHE_SIZE; })
-#define PP_ALLOC_CACHE_REFILL(__pool)	({ (void)__pool; __PP_ALLOC_CACHE_REFILL; })
-struct pp_alloc_cache {
-	u32 count;
-	struct page *cache[__PP_ALLOC_CACHE_SIZE];
-};
+	struct page *cache[PP_ALLOC_CACHE_SIZE];
 #endif	/* !CONFIG_PAGE_POOL_OOB */
+};
 
 struct page_pool_params {
 	unsigned int	flags;
@@ -253,12 +246,6 @@ inline enum dma_data_direction page_pool_get_dma_dir(struct page_pool *pool)
 	return pool->p.dma_dir;
 }
 
-static inline bool page_pool_is_oob(struct page_pool *pool)
-{
-	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
-		pool->p.flags & PP_FLAG_PAGE_OOB;
-}
-
 bool page_pool_return_skb_page(struct page *page);
 
 struct page_pool *page_pool_create(const struct page_pool_params *params);
@@ -272,6 +259,24 @@ void page_pool_use_xdp_mem(struct page_pool *pool, void (*disconnect)(void *),
 void page_pool_release_page(struct page_pool *pool, struct page *page);
 void page_pool_put_page_bulk(struct page_pool *pool, void **data,
 			     int count);
+
+static inline bool page_pool_is_oob(struct page_pool *pool)
+{
+	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
+		pool->p.flags & PP_FLAG_PAGE_OOB;
+}
+
+static inline unsigned int page_pool_cache_size(struct page_pool *pool)
+{
+	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
+		pool->p.flags & PP_FLAG_PAGE_OOB ? pool->p.pool_size : PP_ALLOC_CACHE_SIZE;
+}
+
+static inline unsigned int page_pool_cache_refill(struct page_pool *pool)
+{
+	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
+		pool->p.flags & PP_FLAG_PAGE_OOB ? pool->p.pool_size : PP_ALLOC_CACHE_REFILL;
+}
 #else
 static inline void page_pool_destroy(struct page_pool *pool)
 {
diff --git a/net/core/page_pool.c b/net/core/page_pool.c
index 15bce7a0b2..57aaee2265 100644
--- a/net/core/page_pool.c
+++ b/net/core/page_pool.c
@@ -320,7 +320,7 @@ static struct page *page_pool_refill_alloc_cache(struct page_pool *pool)
 			page = NULL;
 			break;
 		}
-	} while (pool->alloc.count < PP_ALLOC_CACHE_REFILL(pool));
+	} while (pool->alloc.count < page_pool_cache_refill(pool));
 
 	/* Return last page */
 	if (likely(pool->alloc.count > 0)) {
@@ -431,7 +431,7 @@ noinline
 static struct page *__page_pool_alloc_pages_slow(struct page_pool *pool,
 						 gfp_t gfp)
 {
-	const int bulk = PP_ALLOC_CACHE_REFILL(pool);
+	const int bulk = page_pool_cache_refill(pool);
 	unsigned int pp_flags = pool->p.flags;
 	unsigned int pp_order = pool->p.order;
 	struct page *page;
@@ -613,7 +613,7 @@ static bool page_pool_recycle_in_ring(struct page_pool *pool, struct page *page)
 static bool page_pool_recycle_in_cache(struct page *page,
 				       struct page_pool *pool)
 {
-	if (unlikely(pool->alloc.count == PP_ALLOC_CACHE_SIZE(pool))) {
+	if (unlikely(pool->alloc.count == page_pool_cache_size(pool))) {
 		recycle_stat_inc(pool, cache_full);
 		return false;
 	}
-- 
2.30.2

